Lib/python/pybuffer.i

```diff
@@ -12,18 +12,43 @@
  *      }
  */
 
+/* Note that in Py_LIMITED_API case we have no choice, but to use deprecated
+ * functions, as they provides the only way to access buffer data with limited
+ * API, which doesn't include Py_buffer definition. We also disable the
+ * warnings about doing this because they're not useful in our case.
+ */
```

Comment:
So what's the plan here longer term?

Has someone flagged this situation to the Python developers?  Ideally the non-deprecated buffer mechanism gets added to the stable ABI before the deprecated one gets removed from Python.

Reply:
> So what's the plan here longer term?

Sorry, no idea. I need this code to work today and this is the only way to make it work, to the best of my knowledge.

> Has someone flagged this situation to the Python developers? Ideally the non-deprecated buffer mechanism gets added to the stable ABI before the deprecated one gets removed from Python.

Ideally, yes. For now the deprecated functions still exist even in 3.10, in spite of being deprecated since 3.0.

Reply:
I'd love a way to access buffer data via the stable ABI. Probably very much worth bringing up.

Reply:
To be clear, this does work today, when using stable ABI. It just uses deprecated (but still perfectly working) functions for doing it.

Reply:
I am clear on that, but also by merging a new feature we're creating an expectation that we'll support it for some time, so merging a feature which relies on something in the target language API which has been long marked as deprecated is problematic.  It'd be reassuring to know that the Python developers were at least aware of the situation.

Reply:
Turns out I actually have great timing because after checking for it again, the new buffer protocol was [just stabilized](https://bugs.python.org/issue45459) _yesterday_. So we'll be able to start using it from 3.11.

Of course, realistically speaking, it will be quite some time before anybody wants to use stable ABI for 3.11 and later only, so we're going to keep the current code for many years ahead. It's ugly but at least we know that we'll have something to switch over to when it's finally removed (if it ever is, some of the discussions seem to indicate that stable ABI functions will never be removed at all).

I think this should answer your concerns.

Reply:
Mostly.

I think we want to support the new stable buffer API **before** we're forced to by the old one being removed.  Unless SWIG starts to get new releases an order of magnitude or so more frequently users will likely end up stuck with the latest SWIG release not working with the latest Python release in this configuration if we're only reactive here.

It doesn't need to block the merge though.

Reply:
It's not entirely helpful to do reply+"resolve conversation" by the way - means I have to keep clicking to see your replies.

AlsoIME the usual protocol is that it's the reviewer who decides whether the matter is resolved or not...

Reply:
Sorry, I've resolved this one because I thought that the linked issue answers the question about the future and there is really nothing more to do about this today, as we can't rely on everyone having Python 3.11 just yet.

Once 3.11 is released we could at least add another condition and use the new functions in  `Py_LIMITED_API >= 3.11` case, but it would be nice to add CI tests using 3.11 before doing this so, again, I don't think there is anything that can be done today.

In the worst case, if it's really problematic, let's drop this commit. If all the rest is merged and I just have to maintain this one in my branch, I can live with it.

Reply:
> Once 3.11 is released we could at least add another condition and use the new functions in `Py_LIMITED_API >= 3.11` case, but it would be nice to add CI tests using 3.11 before doing this so, again, I don't think there is anything that can be done today.

Python 3.11 has now been released and we now have CI for Python 3.11 (and 3.12) so this could now be done.

I don't think it needs to block the merge, but it'd be great to sort this out soon.